extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateEntityEnhancer.java

```diff
@@ -83,21 +80,63 @@ public void visitEnd() {
         }
 
         private byte[] hibernateEnhancement(final String className, final byte[] originalBytes) {
-            final byte[] enhanced = enhancer.enhance(className, originalBytes);
+            final byte[] enhanced = enhancerHolder.getEnhancer().enhance(className, originalBytes);
             return enhanced == null ? originalBytes : enhanced;
         }
 
     }
 
     public byte[] enhance(String className, byte[] bytes) {
-        DefaultEnhancementContext enhancementContext = new DefaultEnhancementContext() {
-            @Override
-            public ClassLoader getLoadingClassLoader() {
-                return Thread.currentThread().getContextClassLoader();
+        return enhancerHolder.getEnhancer().enhance(className, bytes);
+    }
+
+    private static class EnhancerHolder {
+
+        private volatile Enhancer actualEnhancer;
+
+        public Enhancer getEnhancer() {
+            //Lazily initialized as it's expensive and might not be necessary: these transformations are cacheable.
+            if (actualEnhancer == null) {
+                synchronized (this) {
+                    if (actualEnhancer == null) {
+                        actualEnhancer = PROVIDER.getEnhancer(QuarkusEnhancementContext.INSTANCE, new ThreadsafeLocator());
+                    }
+                }
             }
+            return actualEnhancer;
+        }
+    }
+
+    private static final class ThreadsafeLocator implements EnhancerClassLocator {
+
+        final ThreadLocal<EnhancerClassLocator> localLocator = ThreadLocal
+                .withInitial(() -> ModelTypePool.buildModelTypePool(QuarkusClassFileLocator.INSTANCE,
+                        CORE_POOL));
 
-        };
-        Enhancer enhancer = PROVIDER.getEnhancer(enhancementContext);
-        return enhancer.enhance(className, bytes);
+        @Override
+        public void registerClassNameAndBytes(String s, byte[] bytes) {
+            localLocator.get().registerClassNameAndBytes(s, bytes);
+        }
+
+        @Override
+        public void deregisterClassNameAndBytes(String s) {
+            localLocator.get().deregisterClassNameAndBytes(s);
+        }
+
+        @Override
+        public ClassFileLocator asClassFileLocator() {
+            return localLocator.get().asClassFileLocator();
+        }
+
+        @Override
+        public Resolution describe(String s) {
+            return localLocator.get().describe(s);
+        }
+
+        @Override
+        public void clear() {
+            //not useful
```

Comment:
You could clear the thread-local? Maybe that's not the best place to do this, but it would be nice to at least try... somewhere...

We've been bitten by classloader leaks caused by threadlocals in the past... And it can get quite bad.